packages/commerce-sdk-react/src/hooks/useQuery.ts

```diff
@@ -1,18 +1,42 @@
 /*
- * Copyright (c) 2022, Salesforce, Inc.
+ * Copyright (c) 2023, Salesforce, Inc.
  * All rights reserved.
  * SPDX-License-Identifier: BSD-3-Clause
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 import {useQuery as useReactQuery, UseQueryOptions, QueryKey} from '@tanstack/react-query'
-import useAuthenticatedClient from './useAuthenticatedClient'
-import {IQueryFunction} from './types'
+import {useAuthorizationHeader} from './useAuthorizationHeader'
+import {ApiClients} from './types'
+import {hasAllKeys} from './utils'
 
-export const useQuery = <TData>(
-    queryKey: QueryKey,
-    fn: IQueryFunction<TData>,
-    queryOptions?: UseQueryOptions<TData, Error>
+export const useQuery = <
+    ApiArg extends {headers?: Record<string, string>; parameters?: Record<string, unknown>},
+    Client extends ApiClients[keyof ApiClients],
+    Data,
+    Err,
+    QK extends QueryKey
+>(
+    apiOptions: ApiArg,
+    queryOptions: UseQueryOptions<Data, Err, Data, QK> & {queryKey: QK},
+    hookConfig: {
+        client: Client
+        method: (arg: ApiArg) => Promise<Data>
+        requiredParameters: ReadonlyArray<keyof ApiArg['parameters']>
+        enabled?: boolean
```

Comment:
Intended to be used for extra "default" enabled checks, _in addition to_ the "has all required parameters" logic, so that we don't have to duplicate the required parameter check outside this hook. For example, `useCustomer` might have `enabled: customerType === 'registered'`. The end user can still fully override using `queryOptions.enabled`.